### ISSUE TITLE
Forced into traitor role with no detective when playing with bots

### ISSUE THREAD

**Issue by Mach2Infinity** [BadgerCode/tttdamagelogs#102]
If I have the TTT Damage Logs & RDM Manager addon installed on the server and it's just me and bots, I'm always forced to be the sole traitor with no detective assigned. I first installed the addon when playing with a friend and bots and the TTT role selection would work as normal and be randomised. 

I use the **[TTT Bots 2.0](https://steamcommunity.com/sharedfiles/filedetails/?id=1256344426)** and always add 10 bots so it's an 11 player session. I figured it might have something to do with the **[ttt_detective_min_players](https://www.troubleinterroristtown.com/config/settings/#gameplay)** setting which by default is set to 10, so it being an 11 player game should assign a detective. I also checked the **[config.json](https://github.com/BadgerCode/tttdamagelogs/wiki/Configuring-the-damage-logs)** file but I can't see any code/setting that would affect player roles:

```
{
	"Reports": {
		"MoreReportsPerRound": false,
		"ReportsBeforePlaying": false,
		"NoStaffReports": false
	},
	"DiscordWebhookMode": 0.0,
	"PrivateMessagePrefix": "[RDM Manager]",
	"AbuseMessageMode": 0.0,
	"ForcedLanguage": "",
	"RDM_Manager_Enabled": true,
	"HideDonateButton": false,
	"Autoslay": {
		"Autoslay_CheckCustom": false,
		"DefaultReason11": "Trolling",
		"DefaultReason9": "Prop kill",
		"DefaultReason3": "2x RDM",
		"DefaultReason5": "Mass RDM",
		"DefaultReason6": "Super Mass",
		"ShowRemainingSlays": false,
		"DefaultReason10": "Consistent RDM",
		"DefaultReason8": "Hacking",
		"DefaultReason1": "Random Damage",
		"DefaultReason2": "RDM",
		"DefaultReason12": "Minge",
		"ULX_Autoslay_ForceRole": true,
		"DefaultReason": "Breaking Rules",
		"DefaultReason4": "Attempted Mass",
		"DefaultReason7": "Ghosting",
		"ULX_AutoslayMode": 1.0
	},
	"Use_MySQL": false,
	"LogDays": 61.0,
	"UseWorkshop": true,
	"Commands": {
		"RDM_Manager_Command": "!report",
		"Respond_Command": "!respond"
	},
	"Key": 99.0,
	"Permissions": {
		"owner": {
			"access_level": 4.0,
			"can_access_rdm_manager": true
		},
		"founder": {
			"access_level": 4.0,
			"can_access_rdm_manager": true
		},
		"admin": {
			"access_level": 4.0,
			"can_access_rdm_manager": true
		},
		"superadmin": {
			"access_level": 4.0,
			"can_access_rdm_manager": true
		},
		"operator": {
			"access_level": 3.0,
			"can_access_rdm_manager": false
		},
		"user": {
			"access_level": 2.0,
			"can_access_rdm_manager": false
		}
	},
	"Ban": {
		"DefaultReason6": "Super Mass",
		"DefaultReason7": "Ghosting",
		"DefaultReason8": "Hacking",
		"DefaultReason11": "Random Damage",
		"DefaultReason1": "Random Damage and leave",
		"AllowBanningThruManager": true,
		"DefaultReason2": "RDM and leave",
		"DefaultReason12": "Trolling",
		"DefaultReason3": "2x RDM and leave",
		"DefaultReason10": "Attempted RDM",
		"DefaultReason4": "Attempted Mass and leave",
		"DefaultReason9": "Consistent RDM",
		"DefaultReason5": "Mass RDM"
	}
}
```

### **Attempted Fixes**
1) Added "ttt_detective_min_players 8" (minus quotes) to the server.cfg.
2) Installed v3.8.0 and v3.9.0 separately.
3) I use a 127 item workshop collection, so to rule out if was too many addons I created a test one with just a map, Bots 2.0 and TTTDamagelogs.
4) Deleting cached files on the server using this **[guide](https://nodecraft.com/support/games/gmod/clearing-cached-steam-addons-on-your-garrys-mod-server)**.
5) Deleted cached files on my client.
6) Reinstalled a fresh version of the server.
7) Validated game integrity files on my client.
8) Fully reinstalled the game on my client.
9) Checked AppData in case the were leftover GMod setting files but there's no folder for GMod anyway

There's something I noticed although might be unrelated. When I ran damage logs & RDM manager for the first time, pressing F8 would bring up the logs menu as expected but then afterwards it would open up a sandbox game into a random map on my machine which would disconnect me. I had to unbind F8 through the console to be able to bring up the men again. I also notice I have an extra credit to spare as a traitor as I'm sure it's 2 normally not 3.

If there's anything else I could try or I need to provide further information just let me know.

Thanks.

**Comment by BadgerCode** [comment]
Heyo!
Sorry for the delay.

Quite a weird one, but I think I've found the cause.

It's to do with the slay system.
In order to avoid giving the traitor or detective role to a player who will be slain next round, the addon overrides the TTT function for assigning roles at the start of the round.
https://github.com/BadgerCode/tttdamagelogs/blob/master/lua/damagelogs/server/autoslay.lua#L562

**For some reason, the damage logs code only considers humans for the roles- `player.GetHumans()`**

In the TTT code, it considers all players
https://github.com/Facepunch/garrysmod/blob/b997d469fc90e63dd31d773aac93f31d57dfdaed/garrysmod/gamemodes/terrortown/gamemode/init.lua#L845
`player.GetAll()`


We could try removing this restriction and see what happens.

**Comment by BadgerCode** [comment]
I have created a fix for this- #105 

I've tested it a little with some bots; I spawned as a detective and they were traitors/innocent.
Would you be able to test it a bit further to see if you can spot any problems caused by this?

**Comment by BadgerCode** [comment]
This will be fixed in the next release
